ark/phenotyping/cluster_helpers.py

```diff
@@ -0,0 +1,261 @@
+                                                     range(self.L_, self.K_-1))])
+        self.bestK = np.argmax(self.deltaK) + \
+            self.L_ if self.deltaK.size > 0 else self.L_
+
+    def predict(self):
+        """Predicts on the consensus matrix, for best found cluster number.
+
+        Returns:
+            numpy.ndarray:
+                The consensus matrix prediction for `self.bestK`.
+        """
+        assert self.Mk is not None, "First run fit"
+        return self.cluster_(n_clusters=self.bestK).fit_predict(
+            1-self.Mk[self.bestK-self.L_])
+
+    def predict_data(self, data: pd.DataFrame):
+        """Predicts on the data, for best found cluster number
+
+        Args:
+            data (pandas.DataFrame):
+                `(examples,attributes)` format
+
+        Returns:
+            pandas.DataFrame:
+                The data matrix prediction for `self.bestK`.
+        """
+        assert self.Mk is not None, "First run fit"
+        return self.cluster_(n_clusters=self.bestK).fit_predict(
+            data)
+
+
+class PixieConsensusCluster:
+    def __init__(self, cluster_type: str, input_file: pathlib.Path, columns: List[str],
+                 max_k: int = 20, cap: float = 3):
+        """Constructs a generic ConsensusCluster pipeline object that makes use of
+        Sagovic's implementation of consensus clustering in Python.
+
+        Args:
+            cluster_type (str):
+                The type of data being run through consensus clustering.
+                Must be either `'pixel'` or `'cell'`
+            input_file (pathlib.Path):
+                The average expression values per SOM cluster .csv,
+                computed by `ark.phenotyping.cluster_pixels` or `ark.phenotyping.cluster_cells`
+                depending on the type of data being generated.
+            columns (List[str]):
+                The list of columns to subset the data in `input_file` on for consensus clustering.
+            max_k (int):
+                The number of consensus clusters to use.
+            cap (float):
+                The value to cap the data in `input_file` at after z-score normalization.
+                Data will be within the range `[-cap, cap]`.
+        """
+        # validate the cluster_type provided
+        verify_in_list(
+            provided_cluster_type=cluster_type,
+            supported_cluster_types=['pixel', 'cell']
+        )
+
+        self.cluster_type = cluster_type
+        self.som_col = '%s_som_cluster' % cluster_type
+        self.meta_col = '%s_meta_cluster' % cluster_type
+        self.input_file = input_file
+        self.input_data = pd.read_csv(input_file)
+        self.columns = columns
+        self.max_k = max_k
+        self.cap = cap
+        self.cc = ConsensusCluster(
+            cluster=AgglomerativeClustering,
+            L=max_k,
+            K=max_k,
+            H=10,
+            resample_proportion=0.8
+        )
```

Comment:
Why are `H` and `resample_proportion` default to the values here?

Reply:
It's mostly an attempt to replicate `ConsensusClusterPlus` default args as closely as possible.

`H` is the same as `reps` in `ConsensusClusterPlus`, which is by default set to 10. `resample_proportion` is the same as `pItem` in `ConsensusClusterPlus`, which is by default set to 0.8.

Reply:
Can we add a comment about that, for the future if people are confused where these values came from?